### PR TITLE
modify lro

### PR DESCRIPTION
**Pull request by kazrael2119**
There are two issues introduced when we upgrade the current dependencies and we'll try to fix in this pr.

fixes https://github.com/Azure/autorest.typescript/issues/1511
fixes https://github.com/Azure/autorest.typescript/issues/1516
fixes https://github.com/Azure/autorest.typescript/issues/1544

### CODE REVIEW

**Review by MaryGao — COMMENTED** [review]
Thanks for fixing these upgrading issues.

To pass all cis you have to 

1. Update rlcIntegration/lro.spec.ts also because the upgrading impacted this file
2. Update rlcIntegration/generted and integration/generated folders because the testserver update relevant swagger by command: `rushx generate-swaggers` and `rushx rlc-generate-swaggers`

File: packages/autorest.typescript/test/integration/lro.spec.ts [hunk 1]
```diff
@@ -195,27 +195,14 @@ describe("LROs", () => {
...
-        await client.lROs.beginPost202NoRetry204AndWait(LROOptions);
-        throw new Error("should have thrown instead");
-      } catch (e) {
-        assert.equal(
-          e.message,
-          "Received unexpected HTTP status code 204 while polling. This may indicate a server issue."
-        );
-      }
+      await client.lROs.beginPost202NoRetry204AndWait(LROOptions);
+      check200(lastResponse);
```
Comment (MaryGao) [inline packages/autorest.typescript/test/integration/lro.spec.ts:199]: Why is this check200? According to the name `beginPost202NoRetry204AndWait` It should be 204

File: packages/autorest.typescript/test/integration/lro.spec.ts [hunk 2]
```diff
@@ -195,27 +195,14 @@ describe("LROs", () => {
...
-        await client.lROs.beginDeleteNoHeaderInRetryAndWait(LROOptions);
-        throw new Error("should have thrown instead");
-      } catch (e) {
-        assert.equal(
-          e.message,
-          "Received unexpected HTTP status code 204 while polling. This may indicate a server issue."
-        );
-      }
+      await client.lROs.beginDeleteNoHeaderInRetryAndWait(LROOptions);
+      check200(lastResponse);
```
Comment (MaryGao) [inline packages/autorest.typescript/test/integration/lro.spec.ts:204]: Why is this check200? According to the name beginPost202NoRetry204AndWait It should be 204

File: packages/autorest.typescript/test/integration/lro.spec.ts [hunk 3]
```diff
@@ -250,15 +237,8 @@ describe("LROs", () => {
...
-        await client.lROs.beginDelete202NoRetry204AndWait(LROOptions);
-        throw new Error("should have thrown instead");
-      } catch (e) {
-        assert.equal(
-          e.message,
-          "Received unexpected HTTP status code 204 while polling. This may indicate a server issue."
-        );
-      }
+      await client.lROs.beginDelete202NoRetry204AndWait(LROOptions);
+      check200(lastResponse);
```
Comment (MaryGao) [inline packages/autorest.typescript/test/integration/lro.spec.ts:241]: Again I think It should be 204

File: packages/autorest.typescript/test/rlcIntegration/lroRest.spec.ts [hunk 4]
```diff
@@ -153,44 +153,30 @@ describe("LRO Rest Client", () => {
...
-        await poller.pollUntilDone();
-        throw new Error("should have thrown instead");
-      } catch (e) {
-        assert.equal(
-          e.message,
-          "Received unexpected HTTP status code 204 while polling. This may indicate a server issue."
-        );
-      }
+      const result = await poller.pollUntilDone();
+      assert.equal(result.status, "204");
```
Comment (MaryGao) [inline packages/autorest.typescript/test/rlcIntegration/lroRest.spec.ts:165]: Yes 204 is expected!

File: packages/autorest.typescript/test/rlcIntegration/lroRest.spec.ts [hunk 5]
```diff
@@ -153,44 +153,30 @@ describe("LRO Rest Client", () => {
...
-        throw new Error("should have thrown instead");
-      } catch (e) {
-        assert.equal(
-          e.message,
-          "Received unexpected HTTP status code 204 while polling. This may indicate a server issue."
-        );
-      }
-    });
+      const result = await poller.pollUntilDone();
+      assert.equal(result.status, "204");
```
Comment (MaryGao) [inline packages/autorest.typescript/test/rlcIntegration/lroRest.spec.ts:178]: +1

File: packages/autorest.typescript/test/rlcIntegration/lroRest.spec.ts [hunk 6]
```diff
@@ -257,24 +243,17 @@ describe("LRO Rest Client", () => {
...
-        await poller.pollUntilDone();
-        throw new Error("should have thrown instead");
-      } catch (e) {
-        assert.equal(
-          e.message,
-          "Received unexpected HTTP status code 204 while polling. This may indicate a server issue."
-        );
-      }
+      const result = await poller.pollUntilDone();
+      assert.equal(result.status, "204");
```
Comment (MaryGao) [inline packages/autorest.typescript/test/rlcIntegration/lroRest.spec.ts:256]: +1

File: packages/autorest.typescript/test/integration/lro.spec.ts (hunk 1 above)
Comment (deyaaeldeen) [reply packages/autorest.typescript/test/integration/lro.spec.ts:199]: I think it should be a 204, see https://github.com/Azure/azure-sdk-for-js/blob/1f9325c66677b85c96b3119a3a17dfe093806d31/sdk/core/core-lro/test/engine.spec.ts#L67-L96

**Review by deyaaeldeen — COMMENTED** [review]
Looks good. I can approve once CI is green.

File: packages/autorest.typescript/test/rlcIntegration/lroRest.spec.ts [hunk 7]
```diff
@@ -995,7 +937,7 @@ describe("LRO Rest Client", () => {
         await poller.pollUntilDone();
         assert.fail("Scenario should throw");
       } catch (error) {
-        assert.equal(error.statusCode, 400);
+        assert.equal(error.message, "status.toLowerCase is not a function");
```
Comment (MaryGao) [inline packages/autorest.typescript/test/rlcIntegration/lroRest.spec.ts:940]: @deyaaeldeen Could you refine the error message? The below one doesn't make sense from client side.

```
status.toLowerCase is not a function
```

Comment (deyaaeldeen) [reply packages/autorest.typescript/test/rlcIntegration/lroRest.spec.ts:940]: This has been addressed in https://github.com/Azure/azure-sdk-for-js/pull/23296. The following message is raised instead:
> Polling was unsuccessful. Expected status to have a string value or no value but it has instead: ${status}. This doesn't necessarily indicate the operation has failed. Check your Azure subscription or resource status for more information.

**Comment by joheredi** [comment]
Thanks for your contribution! I see there are changes to the generated clients, however, I couldn't find any changes to the generator. Are all these manual changes?

**Comment by deyaaeldeen** [comment]
@azure/core-lro@2.4.0 has been released, could you please upgrade to that version and update the code generator to turn on the `resolveOnUnsuccessful` parameter in `LroEngineOptions` for RLC clients?

**Comment by MaryGao** [comment]
> Thanks for your contribution! I see there are changes to the generated clients, however, I couldn't find any changes to the generator. Are all these manual chang

**Comment by MaryGao** [comment]
Close this one and reopen the pr to the main branch

**Pull request closed without merging**